File: src/main/java/gg/mineral/practice/util/collection/Registry.java
package gg.mineral.practice.util.collection;

import java.util.Collection;
import java.util.Map;
import java.util.function.Function;

import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class Registry<T, Q> {
    Map<Q, T> keyValueMap = new Object2ObjectOpenHashMap<>();
    final Function<T, Q> function;

    public Collection<T> getRegisteredObjects() {
        return keyValueMap.values();
    }

    public void clear() {
        keyValueMap.clear();
    }

    public T get(String key) {
        return keyValueMap.get(key);
    }

    public void put(T value) {
        keyValueMap.put(function.apply(value), value);
    }

    public void unregister(T value) {
        keyValueMap.remove(function.apply(value));
    }
}

File: src/main/kotlin/gg/mineral/practice/util/collection/Registry.kt
package gg.mineral.practice.util.collection

import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import java.util.function.Function

class Registry<T, Q>(private val function: Function<T, Q>) {
    private val keyValueMap: MutableMap<Q, T> = Object2ObjectOpenHashMap()

    val registeredObjects: Collection<T>
        get() = keyValueMap.values

    fun iterator(): Iterator<Map.Entry<Q, T>> = keyValueMap.entries.iterator()

    fun clear() = keyValueMap.clear()

    fun get(key: Q) = keyValueMap[key]

    fun put(value: T) {
        keyValueMap[function.apply(value)] = value
    }

    fun unregister(value: T) {
        keyValueMap.remove(function.apply(value))
    }
}
